### ISSUE TITLE
question(format): Looking for some advice on an idea I've been playing around with.

### ISSUE THREAD

**Issue by ttowncompiled** [systemjs/systemjs#711]
I've been learning about systemjs for a little while now and I'm really liking the structure of the format. I did some tinkering with different dependency graphs and I came up with a few ideas for the format that I was interested to get your opinion on.

My team has been successfully using traceur and the output of traceur would look something like so,
```JavaScript
System.register(['A', 'B'], function($export) {
    var _A_, _B_;
    return {
        setters: [
            function(_m_) {
                _A_ = _m_.A;
            },
            function(_m_) {
                _B_ = _m_.B;
                $export('B', _m_.B);
            }
        ],
        execute: function() {
            $export('C', (function() {
                function C() {
                }
                return C;
            })() );
            var a = new _A_();
            var b = new _B_();
            var four = 4;
            $export('four', four);
        }
    };
});
```
As I understand it (please correct me where I'm wrong). After the module calls ```System.register```, the module is ```declared``` which allows the variables ```_A_, _B_``` to be hoisted and defined within the scope of the anonymous function passed into the register call.

Since it has dependencies, both of those dependencies, ```'A', 'B'``` respectively, will need to be loaded. Those modules will be loaded and each passed a weak reference back to ```setters``` declared here.

Let's say those modules have no dependencies. Both will be ```declared``` and ```executed```. When they are ```executed```, each will ```export``` their respective class (```A``` for dependency A and ```B``` for dependency B).

When each of those calls ```export``` they will also call their respective ```setter``` for the module above to dynamically update its references, ```_A_, _B_``` respectively. This will set ```_A_ = A;```, ```_B_ = B;``` and will also cause this module to ```export B```.

Once those two modules have been ```executed```, this module may be ```executed```. This module will ```export C```, create variables ```a, b```, and also ```export four```.

The assumption here being that for any module, we can aptly divide that module up into three parts: declaration, imports/exports, and execution. Which of course, we know that every module can be broken up this way.

**TLDR;**
My suggestion is to modify the above to the following,
```JavaScript
System.register(['A', 'B'], function($export) {
    var _A_, _B_;
    $export('C', (function() {
        function C() {
        }
        return C;
    })() );        
    return {
        setters: [
            function(_m_) {
                _A_ = _m_;
            },
            function(_m_) {
                _B_ = _m_;
            }
        ],
        execute: function() {
            $export('B', _B_.B);
            var a = new _A_();
            var b = new _B_();
            var four = 4;
            $export('four', four);
        }
    };
});
```
I believe this helps for the following reasons.

First, this moves all values that are passed by reference to the declaration section such as **classes** and **functions**. These do not need to be evaluated so exporting them there won't generate errors. If they need to be extended, they can of course be moved to the ```execute``` block. The benefit of this is that circular references are now much easier to evaluate. If two modules depend on one another for exported classes or functions, those modules may be executed in any order without consequence since the references would have been defined the moment the module was loaded. This assumes that all other dependencies have been met first.

Secondly, this changes the spec to disallow assigning references of module properties to a local variable and only allows assigning module references to local variables. This removes the need to ever call a setter more than once. Since the reference to the module will never change, once that module has been ```declared``` and ```executed```, it is guaranteed to not break its importers.

Thirdly, this moves transitive exports to the execute section. Currently, if modules transitively export references, they go through the process of calling the setter of their importers multiple times with an ```undefined``` value because the reference their exporting wasn't defined yet. By moving these to the ```execute``` block, it is guaranteed to export the actual value.

Lastly, this opens the doors to advancements in the future. TypeScript provides the means to adequately determine if exports need to be ```declared``` or ```executed```. If we can determine that an export does not need to ```executed```, then we could do things such as move transitive exports to the ```declaration``` section. We could also move super and sub classes to the declaration section.

The idea being to divide modules into two sections. The ```declaration``` section would contain all code from the module that does not evaluate a reference. The ```execute``` section would contain all code that does need to evaluate a reference.

**Comment by guybedford** [comment]
We do this for functions, but classes have what is called a temporal dead zone, meaning that they are only defined on execution (due to the fact that extends can contain arbitrary expressions). As a result classes don't work nicely with circular references, but that is just the way JS is specified. Let me know if that answers your questions here.

**Comment by ttowncompiled** [comment]
I wasn't aware of TDZ. Thanks! That actually helps a lot.

**Comment by guybedford** [comment]
Sure!
